### DISCUSSION TITLE
Move to new Golang build process.

### DISCUSSION THREAD

**Discussion by Edu4rdSHL** [BlackArch/blackarch#3083] [Ideas]
Hello,

I have created https://github.com/BlackArch/blackarch-pkgbuilds/pull/10 in order to switch from the old Golang build process to a new one that brings a lot of advantages and is technically superior and simple.

## Why

As mentioned previously, there are several advantages:

1. Option `-trimpath` to go build makes them system independent and help creating [reproducible builds](https://reproducible-builds.org/). The option instructs the compiler to remove all file system paths from the compiled executable. As a result, you will see module paths in stack traces instead of absolute paths of the build environment. Also prevents issues like #3082.
2. Option `-buildmode=pie` build executables in [Position-independent Code](https://en.wikipedia.org/wiki/Position-independent_code) mode.
3. Option `-mod=readonly ` resolve module imports, but fail if they must be reconfigured, means that it prevents the build process from modifying the go.mod file.
4. Option `-modcacherw` instructs the go command to leave newly-created directories in the module cache at their default permissions rather than making them read-only.
5. Option `-ldflags "-s -w"` strip [DWARF](https://en.wikipedia.org/wiki/DWARF), symbol table and debug info, it leads to ~25% binary size decrease or more in most cases.

Additionally to the previous mentioned benefits, this build mode will rarely fails.

## TODO

If this proposition is accepted, then I will take care of migrating all packages that depend on Go to the new build model. Of course, help is welcome.

Regards,
Ed

**Comment by noptrix** [comment]
@Edu4rdSHL this is brilliant. thanks a lot!

**Comment by Edu4rdSHL** [comment]
TZAG,

After two days of hard work I have finished the migration of all the Go packages to the new build process as proposed here. The summary can be found in this [attached file](https://github.com/BlackArch/blackarch/files/6381841/golang.TXT) with the following conventions:


* PKGBUILD+ = Package was successfully migrated. [dnsgrep](https://github.com/BlackArch/blackarch/blob/1c28e5a147a8d9cf5528b6a77a4184d67697c0e6/packages/dnsgrep/PKGBUILD#L28-L51) and [raven](https://github.com/BlackArch/blackarch/blob/1c28e5a147a8d9cf5528b6a77a4184d67697c0e6/packages/raven/PKGBUILD#L25-L44) are two special cases where I have to make an "horrible hack" copying `/usr/lib/go` to `$srcdir`, perform other operations and change `GOROOT` while building. As it doesn't affect the end user, not even the build system, I didn't see any problem there.
* PKGBUILD- = Package was not migrated. Reason is put there as well.

Regards,
Ed
